add AbstractEventPublisher

TrueSamedi opened the pull request:
add LikeEvent
add LikeEventPublisher
add RedisConfig
add SerializeJsonException
fixed application.yaml
fixed LikeController
fixed LikeDto
fixed LikeMapper
fixed LikeService
fixed LikeServiceImpl
fixed LikeServiceTest

TrueSamedi committed:
add AbstractEventPublisher
add LikeEvent
add LikeEventPublisher
add RedisConfig
add SerializeJsonException
fixed application.yaml
fixed LikeController
fixed LikeDto
fixed LikeMapper
fixed LikeService
fixed LikeServiceImpl
fixed LikeServiceTest

PavelAP96 commented on src/main/java/faang/school/postservice/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,46 @@
+    @Value("${spring.data.redis.port}")
+    private int post;
+    @Value("${spring.data.redis.channels.like_channel.name}")
+    private String likeEvent;
```
название не совсем соответствует тому, что в это переменной будет храниться
скорее это likeEventChannel

PavelAP96 commented on src/main/java/faang/school/postservice/exception/SerializeJsonException.java:
```diff
@@ -0,0 +1,7 @@
+package faang.school.postservice.exception;
+
+public class SerializeJsonException extends RuntimeException {
```
уверен, что надо создавать новое исключение?
на сколько помню есть уже существующие исключения, подходящие тебе

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,27 @@
+
+    private final RedisTemplate<String, Object> redisTemplate;
+    private final ObjectMapper objectMapper;
+    private final String topicChannelName;
```
в абстрактном классе хранятся только общие параметры, которые одинаковы для всех его наследников
когда у тебя добавятся сюда еще события, которые будут расширять этот класс, то начнутся проблемы

вынеси эту переменную в LikeEventPublisher и передавай ее в метод publishInTopic

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,27 @@
+    private final ObjectMapper objectMapper;
+    private final String topicChannelName;
+
+    protected void publishInTopic(T event) {
```
я бы просто назвал send или convertAndSend
итак понятно, что он будет пулять в топик

даже если оставлять твое название, то предлог In не корректно тут использовать, должен использоваться to

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,27 @@
+    private final String topicChannelName;
+
+    protected void publishInTopic(T event) {
+        String json;
+        try {
+            json = objectMapper.writeValueAsString(event);
+        } catch (JsonProcessingException e) {
+            throw new SerializeJsonException("Cannot serialize event to json");
+        }
+        redisTemplate.convertAndSend(topicChannelName, json);
```
а почему ты не хочешь все это засунуть в блок try?
будет лучше код смотреться
```
 try {
    String json = objectMapper.writeValueAsString(event);
    redisTemplate.convertAndSend(topicChannelName, json);
} catch (JsonProcessingException e) {
    log.error("Json processing exception", e);
    throw new RuntimeException("Json processing exception");
}
```

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/AbstractEventPublisher.java:
```diff
@@ -0,0 +1,27 @@
+
+@Component
+@RequiredArgsConstructor
+public abstract class AbstractEventPublisher<T> {
```
предлагаю еще сюда добавить абстрактный метод publish, который должны будут реализовывать все его наследники, подумай, что в него должно идти на вход, он будет только вызывать convertAndSend и передавать нужные параметры в него

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/LikeEventPublisher.java:
```diff
@@ -0,0 +1,32 @@
+    @Autowired
+    public LikeEventPublisher(RedisTemplate<String, Object> redisTemplate,
+                              ObjectMapper objectMapper,
+                              @Value("${spring.data.redis.channels.like_channel}") String likeChannelName,
```
я уже писал тебе в абстрактном классе
пусть эта переменная будет объявлена в этом классе

PavelAP96 commented on src/main/java/faang/school/postservice/publisher/LikeEventPublisher.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public void publish(Like like) {
+        LikeEvent event = likeMapper.toEventDto(like);
+        event.setCreatedAt(LocalDateTime.now());
```
не совсем логично делать это здесь
этот метод у нас отвечает только за публикацию, все преобразования должны происходить до этого и ему на вход должен передаваться уже готовый объект LikeEvent

соответственно надо перенести эту логику в LikeService

TrueSamedi committed:
delete SerializeJsonException
fixed AbstractEventPublisher
fixed LikeEventPublisher
fixed LikeServiceImpl
fixed RedisConfig
